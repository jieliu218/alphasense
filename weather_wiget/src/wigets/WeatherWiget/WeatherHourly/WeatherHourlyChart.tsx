import React, { FC } from "react";
import { ApexOptions } from "apexcharts";
import {
  Box,
  useTheme,
  BoxProps,
  Divider,
  Stack,
  Typography,
} from "@mui/material";
import Chart, { useChart } from "../../../components/Chart";
import { bgBlur, textBlur } from "../../../theme/css";
import { fDateTime } from "../../../utils/formatTime";
import { fNumber } from "../../../utils/formatNumber";
import { Maybe } from "../../../__generated__/graphql";

interface WeatherHourlyChartProps extends BoxProps {
  title: string;
  icon: React.ReactNode;
  time: string[] | undefined;
  unit: Maybe<string> | undefined;
  chart: {
    colors?: string[];
    series: number[];
    options?: ApexOptions;
  };
}

const WeatherHourlyChart: FC<WeatherHourlyChartProps> = ({
  title,
  icon,
  time,
  unit,
  chart,
  sx,
  ...other
}) => {
  const theme = useTheme();

  const {
    colors = [theme.palette.primary.light, theme.palette.primary.main],
    series,
    options,
  } = chart;

  const chartOptions = useChart({
    colors: [colors[1]],
    fill: {
      type: "gradient",
      gradient: {
        colorStops: [
          // @ts-expect-error: ApexCharts type error
          { offsets: 0, color: colors[0] },
          { offsets: 100, color: colors[1] },
        ],
      },
    },
    chart: {
      animations: {
        enabled: true,
      },
      sparkline: {
        enabled: true,
      },
    },
    xaxis: {
      categories: (time || []).map((t) => fDateTime(t, "dd MMM p")),
    },
    tooltip: {
      x: {
        show: true,
      },
      y: {
        formatter: (value: number) => fNumber(value) + unit,
        title: {
          formatter: () => "",
        },
      },
      marker: {
        show: false,
      },
    },
    ...options,
  });

  const textColor = textBlur(theme.palette.common.white, 0.7);

  return (
    <Box
      sx={{
        display: "flex",
        flexDirection: "column",
        bgcolor: bgBlur({ color: theme.palette.common.white, opacity: 0.2 }),
        p: 2,
        borderRadius: 2,
        ...sx,
      }}
      {...other}
    >
      <Stack
        direction="row"
        justifyContent="flex-start"
        alignItems="center"
        spacing={0.25}
      >
        {icon}
        <Typography
          gutterBottom
          variant="body1"
          component="span"
          color={textColor}
        >
          {title}
        </Typography>
      </Stack>
      <Divider sx={{ mt: 2, mb: 2 }} />
      <Chart
        type="line"
        series={[{ data: series as number[] }]}
        options={chartOptions}
        width={"100%"}
        height={64}
      />
    </Box>
  );
};

export default WeatherHourlyChart;
