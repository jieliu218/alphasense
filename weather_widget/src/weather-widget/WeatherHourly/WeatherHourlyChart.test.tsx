import { render, screen } from "@testing-library/react";
import WeatherHourlyChart from "./WeatherHourlyChart";

jest.mock("@mui/material/styles", () => ({
  ...jest.requireActual("@mui/material/styles"),
  useTheme: () => ({
    palette: {
      primary: {
        light: "#fff",
        main: "#000",
      },
      common: {
        white: "#fff",
      },
    },
  }),
}));

jest.mock("../../components/Chart", () => ({
  __esModule: true,
  default: () => <div>MockedChart</div>,
  useChart: jest.fn().mockImplementation((options) => options),
}));

jest.mock("../../utils/formatTime", () => ({
  fDateTime: jest.fn().mockReturnValue("Formatted Date"),
}));

describe("WeatherHourlyChart", () => {
  it("renders correctly with given props", () => {
    const mockProps = {
      title: "Test Chart",
      icon: <span>Icon</span>,
      time: ["2022-01-01T00:00:00Z", "2022-01-01T01:00:00Z"],
      unit: "Â°C",
      chart: {
        series: [1, 2],
      },
    };

    const { asFragment } = render(<WeatherHourlyChart {...mockProps} />);

    expect(screen.getByText("Test Chart")).toBeInTheDocument();

    expect(screen.getByText("Icon")).toBeInTheDocument();

    expect(screen.getByText("MockedChart")).toBeInTheDocument();

    expect(asFragment()).toMatchSnapshot();
  });
});
