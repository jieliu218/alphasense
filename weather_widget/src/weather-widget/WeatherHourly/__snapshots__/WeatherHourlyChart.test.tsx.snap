// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherHourlyChart renders correctly with given props 1`] = `
<DocumentFragment>
  <div
    class="MuiBox-root css-1acq8i5"
  >
    <div
      class="MuiStack-root css-qp5e3-MuiStack-root"
    >
      <span>
        Icon
      </span>
      <span
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom css-zt1oku-MuiTypography-root"
      >
        Test Chart
      </span>
    </div>
    <hr
      class="MuiDivider-root MuiDivider-fullWidth css-ctpjxq-MuiDivider-root"
    />
    <div>
      MockedChart
    </div>
  </div>
</DocumentFragment>
`;
